### PR TITLE
[FLINK-25600][table-planner] Support new statement set syntax in sql client and update docs

### PR DESCRIPTION
**Pull request by wenlong88**
## What is the purpose of the change
this pr is a follow up work of FLINK-25392, aims at adding support new statement set syntax in sql client and update related docs.

## Brief change log
The new statement set will contain multi line statement which ends with `;` in order to support such case, I do some refactor on the sql command parser
     1. make extended command parsers support command end with ';'
     2. support partial parsing in sql parser, converting incomplete statement to StagingOperation
     3. LinerParser in sql client marks a statement as finished only when parsing succeed.

## Verifying this change

This change added tests and can be verified as follows:
  - *Added test on partial parsing*
  - *Added it test on new statement set*
  - *Other refactor changes can be verified in existing tests. *

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: docs/content.zh/docs/dev/table/sqlClient.md [hunk 1]
```diff
@@ -459,7 +459,8 @@ of executing multiple queries.
 
 #### Syntax
```
Comment (godfreyhe) [inline docs/content.zh/docs/dev/table/sqlClient.md:460]: declare that `BEGIN STATEMENT SET` is deprecated.

File: docs/content/docs/dev/table/sql/explain.md [hunk 2]
```diff
@@ -288,4 +288,4 @@ EXPLAIN [([ExplainDetail[, ExplainDetail]*]) | PLAN FOR] <query_statement_or_ins
 ```
 
 For query syntax, please refer to [Queries]({{< ref "docs/dev/table/sql/queries/overview" >}}) page.
-For INSERT, please refer to [INSERT]({{< ref "docs/dev/table/sql/insert" >}}) page.
+For INSERT, the statement can be either a single insert or a statement set, please refer to [INSERT]({{< ref "docs/dev/table/sql/insert" >}}) page.
```
[the remark this answers is not in the document]
Comment (wenlong88) [reply docs/content/docs/dev/table/sql/explain.md:291]: the syntax of statement set is present at insert.md, I think we don't need to add syntax here, just like single insert.

File: flink-table/flink-sql-client/src/test/resources/sql_multi/statement_set.q [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+!output
+[ERROR] Could not execute SQL statement. Reason:
+org.apache.flink.sql.parser.impl.ParseException: Encountered "END" at line 2, column 1.
+Was expecting one of:
+    "INSERT" ...
+    "UPSERT" ...
+
+!error
+
+EXECUTE STATEMENT SET BEGIN
```
[the remark this answers is not in the document]
Comment (wenlong88) [reply flink-table/flink-sql-client/src/test/resources/sql_multi/statement_set.q:172]: is it necessary? line parser would not return when there is no ; at the end of a line.

**Comment by wenlong88** [comment]
@flinkbot run azure

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java [hunk 4]
```diff
@@ -130,4 +130,12 @@
                     .defaultValue(true)
                     .withDescription(
                             "When it is true, the optimizer will simplify the operator name with id and type of ExecNode and keep detail in description. Default value is true.");
+
+    public static final ConfigOption<Boolean> TABLE_OPTIMIZER_ALLOW_PARTIAL_PARSE =
```
[the remark this answers is not in the document]
Comment (wenlong88) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java]: I moved it to TableConfigOptions

File: docs/content/docs/dev/table/sql/insert.md [hunk 5]
```diff
@@ -271,4 +276,37 @@ INSERT INTO students
...
+```sql
+EXECUTE STATEMENT SET
+BEGIN
+insert_spec;
+...
+insert_spec;
+END;
+
+insert_spec:
+   {insert_from_select}|{insert_from_values}
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply docs/content/docs/dev/table/sql/insert.md:293]: I do not see the variable definition

File: docs/content/docs/dev/table/sql/explain.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (godfreyhe) [reply docs/content/docs/dev/table/sql/explain.md:291]: Although read the  document of `insert.md`,  users may do not know the syntax of `explain statement set begin ... end`

File: flink-table/flink-sql-client/src/test/resources/sql/set.q [hunk 6]
```diff
@@ -48,6 +48,7 @@ set;
 'pipeline.jars' = ''
 'rest.port' = '$VAR_REST_PORT'
 'table.exec.sink.legacy-cast-behaviour' = 'DISABLED'
+'table.optimizer.allow-partial-parse' = 'true'
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-sql-client/src/test/resources/sql/set.q]: I do not think we should support it now because table env api does not support parsing multiple line now.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java [hunk 7]
```diff
@@ -95,12 +105,24 @@ public ParserImpl(
...
+                    SqlToOperationConverter.convert(planner, catalogManager, parsed.get(0))
+                            .orElseThrow(
+                                    () -> new TableException("Unsupported query: " + statement)));
+        } catch (SqlParserException parserException) {
+            if (config.getConfiguration().getBoolean(TABLE_PARSER_ALLOW_PARTIAL_PARSE)
+                    && parserException.getMessage().contains("Encountered \"<EOF>\"")) {
+                return Collections.singletonList(new StagingOperation(statement, parserException));
+            } else {
+                throw parserException;
+            }
```
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java:124]: I would rather implement this differently: More than having an operation here that wraps the parsing exception etc, which then is used to properly handle EOF by the client, I would instead have a new method in the parser that helps with this "interactive parsing" mode.

For example, it could be some method like `interactiveParse` which returns a union type: either the operation or the parse failure. No need to use the operation stack for that, as it's really counterintuitive. This new method will also allow you to get rid of this internal config option and simplify the implementation overrall.

Also, the check based on string seems a bit fragile. Can we have a more robust check of some kind? Is there some sub exception that calcite uses?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java [hunk 8]
```diff
@@ -57,6 +58,15 @@ public SqlNode parse(String sql) {
         }
     }
 
+    public SqlNodeList parseSqlList(String sql) {
```
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java:61]: Add some javadoc and also add a `throws` to specify you throw the flink's exception, rather than the calcite one

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/ClearOperationParseStrategy.java [hunk 9]
```diff
@@ -29,7 +29,7 @@
     static final ClearOperationParseStrategy INSTANCE = new ClearOperationParseStrategy();
 
     private ClearOperationParseStrategy() {
-        super(Pattern.compile("CLEAR", DEFAULT_PATTERN_FLAGS));
```
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/ClearOperationParseStrategy.java:32]: Why this change?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/parse/SetOperationParseStrategyTest.java [hunk 10]
```diff
@@ -35,6 +35,13 @@ public void testMatches() {
         assertTrue(
                 SetOperationParseStrategy.INSTANCE.match(
                         "SET table.local-time-zone = 'Europe/Berlin'"));
+        assertTrue(SetOperationParseStrategy.INSTANCE.match("SET;"));
```
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/parse/SetOperationParseStrategyTest.java:38]: Use assertj

**Review by slinkydeveloper — CHANGES_REQUESTED** [review]
I think we should reiterate over the changes in the `Parser`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/ClearOperationParseStrategy.java (hunk 9 above)
Comment (wenlong88) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/ClearOperationParseStrategy.java:32]: it is because that in this pr, we didn't truncate ';' in sql client, parser need to support commands in both 'clear' and 'clear;'

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java (hunk 7 above)
Comment (wenlong88) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java]: thanks for the advise, I am also thinking about how to get rid of the internal config based on the Godfreyhe 's comment. I like the idea adding a new function, but I would like to keep the StagingOperation instead of a union type, because it make it easier to transporting the result between different components in parser/client. 

I am afraid that we can only check based on string, because the calcite parser only format the message to describe the error, there is no sub exception.

WDYT?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java [hunk 11]
```diff
@@ -76,16 +81,42 @@ public ParserImpl(
...
+     * @return parsed operations.
+     */
+    @Override
+    public List<Operation> parse(String statement) {
+        return parseInternal(statement, false);
+    }
+
+    /**
+     * Parses a given statement interactively to {@link Operation}, used in sql client. Different
+     * from {@link #parse(String)}, it returns a {@link StagingOperation} when given an incomplete
```
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java:98]: Please don't use `StagingOperation`, as it's confusing the fact that this is an implementation of an `Operation`. Create an appropriate new union type that either has the `List<Operation>` in case the parsing is complete, or has only `SqlParseException` in case of failure, or some flag to say `expected <EOF>`. e.g. a type like:


`InteractiveParseResult`:
* `isParsingCompleted()` and `getParseResult()` in case parsing is complete
* `isParsingFailured()` and `getParseError()` in case parsing is complete but there was an error
* `isParsingIncomplete()` if the input string was an incomplete statement

Of course choose the names you think fit most, those are just a suggestion of how this type should look like

Comment (wenlong88) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java]: hi, I have investigated again. currently we don't need the new function: when the statement is incomplete, parser throws SqlParserEOFException, and we can just check the exception in sql client. 
WDYT?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java [hunk 12]
```diff
@@ -47,12 +49,36 @@ public CalciteParser(SqlParser.Config config) {
      * @param sql a sql string to parse
      * @return a parsed sql node
      * @throws SqlParserException if an exception is thrown when parsing the statement
+     * @throws SqlParserEOFException if the statement is incomplete
      */
-    public SqlNode parse(String sql) {
+    public SqlNode parse(String sql) throws SqlParserException {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java:54]: need not declare the runtime exception

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/Parser.java [hunk 13]
```diff
@@ -44,9 +45,9 @@
      *
      * @param statement the SQL statement to evaluate
      * @return parsed queries as trees of relational {@link Operation}s
-     * @throws org.apache.flink.table.api.SqlParserException when failed to parse the statement
+     * @throws SqlParserException when failed to parse the statement
      */
-    List<Operation> parse(String statement);
+    List<Operation> parse(String statement) throws SqlParserException;
```
Comment (godfreyhe) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/Parser.java:50]: ditto

**Review by godfreyhe — COMMENTED** [review]
Thanks for the update, LGTM overall exception two minor comments

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java [hunk 14]
```diff
@@ -264,25 +270,43 @@ public boolean executeInitialization(String content) {
...
 
-            String line;
+            Optional<Operation> parsedOperation = Optional.empty();
             try {
-                line = lineReader.readLine(prompt, null, inputTransformer, null);
+                String line = lineReader.readLine(prompt, null, inputTransformer, null);
+                if (line.trim().isEmpty()) {
+                    continue;
+                }
+                parsedOperation = parser.getParsedOperation();
```
Comment (slinkydeveloper) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java:299]: I had a bit of hard time to understand how state propagation works here. I think we can do better and make it more clear: You can make`SqlMultiLineParser` stateless, and propagate every parsing result wrapping `parseException` and `parsedOperation` within `SqlArgumentList` (perhaps also rename this class), and then here you can access the last parsed `SqlArgumentList` using `lineReader.getParsedLine()`.

With this you don't need to check any correctness of state propagation (with the below checkArgument) and you don't need to keep the latest command line. And it makes the code overrall simpler to understand.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/SqlMultiLineParser.java [hunk 15]
```diff
@@ -31,18 +37,61 @@
  */
 public class SqlMultiLineParser extends DefaultParser {
```
Comment (slinkydeveloper) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/SqlMultiLineParser.java:38]: This can be package private

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/Parser.java [hunk 16]
```diff
@@ -44,7 +45,7 @@
      *
      * @param statement the SQL statement to evaluate
      * @return parsed queries as trees of relational {@link Operation}s
-     * @throws org.apache.flink.table.api.SqlParserException when failed to parse the statement
+     * @throws SqlParserException when failed to parse the statement
```
[the remark this answers is not in the document]
Comment (wenlong88) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/Parser.java:48]: I would like to revert the change in docs here, we could discuss later if you think it is necessary to delete the throws or add throw in signature of the method

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java [hunk 17]
```diff
@@ -47,12 +49,36 @@ public CalciteParser(SqlParser.Config config) {
...
+     * @throws SqlParserException if an exception is thrown when parsing the statement
+     * @throws SqlParserEOFException if the statement is incomplete
+     */
+    public SqlNodeList parseSqlList(String sql) {
+        try {
+            SqlParser parser = SqlParser.create(sql, config);
+            return parser.parseStmtList();
+        } catch (SqlParseException e) {
+            if (e.getMessage().contains("Encountered \"<EOF>\"")) {
+                throw new SqlParserEOFException(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (wenlong88) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java:80]: personally I like explicit subclass better, it can make code easier to understand.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/SqlMultiLineParser.java (hunk 15 above)
Comment (slinkydeveloper) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/SqlMultiLineParser.java:38]: I think we should by default try to hide as much as possible. In this case, this class is not used outside this package (and doesn't really have any specific utility outside sql client anyway), so we should make it package private

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java:80]: That's personal preference and it's not a blocking issue for me. But i can see how we might want in future to catch other parsing errors (also see in context of https://cwiki.apache.org/confluence/display/FLINK/FLIP-189%3A+SQL+Client+Usability+Improvements) and one class per error type is overkill and it's going to explode. The point is that if what you need is just to match the specific error type variant, but the variants themselves don't have any other useful error data, then you don't subclasses.

[the remark this answers is not in the document]
Comment (wenlong88) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java:80]: so can we keep the class for now and do some necessary refactor in the future when necessary?

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/SqlMultiLineParser.java (hunk 15 above)
Comment (wenlong88) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/SqlMultiLineParser.java:38]: if you really think it is necessary, I can remove the `public`

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java (hunk 14 above)
Comment (slinkydeveloper) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java:299]: I see, then let's keep it as it is. Perhaps just comment the method `getParsedOperation` and a bit the code here to explain the relationship and the state sharing between these two objects

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java:80]: What are the implications if we do this now vs later? Also look at this point in particular: https://cwiki.apache.org/confluence/display/FLINK/FLIP-189%3A+SQL+Client+Usability+Improvements#FLIP189:SQLClientUsabilityImprovements-ImplementationDetails this talks about how in future we might want to extend `SqlMultiLineParser` to catch more parsing errors to leverage other jline features. Having a single sub exception for all of these is going to pollute our classpath.

[the remark this answers is not in the document]
Comment (wenlong88) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java:80]: I think you may misunderstand the FLIP,  the FLIP want to add some quick check on the statement provided by jline, instead of using Flink Parser. I don't see the need to introduce Exception with ErrorType for now actually. If in the future we really need to introduce more exception, we can remove the EOFException, replace with the new one.

**Review by slinkydeveloper — APPROVED** [review]

**Review by godfreyhe — APPROVED** [review]

**Comment by godfreyhe** [comment]
waiting azure test result
